code/src/pgsql_io/pgsql_io.ml

```diff
@@ -842,6 +842,148 @@ module Prepared_stmt = struct
+  }
+
+  type server_first = {
+    nonce : string;
+    salt : string;
+    iter : int;
+    message : string;
+  }
+
+  type client_final = {
+    server_key : string;
+    auth_message : string;
+    message : string;
+  }
+
+  (* Notation and function naming come from RFC 5802, to avoid
+     messing up the implementation:
+        https://www.rfc-editor.org/rfc/rfc5802#section-2.2 *)
+  let hmac key msg =
+    let hash = Cryptokit.MAC.hmac_sha256 key in
+    Cryptokit.hash_string hash msg
+
+  let h hf msg = Cryptokit.hash_string hf msg
+
+  let hi str salt iter =
+    let ui = hmac str (salt ^ "\x00\x00\x00\x01") in
+    let u = Bytes.of_string ui in
+    let rec loop it ui =
+      match it with
+      | 0 -> String.of_bytes u
+      | k ->
+          let ui = hmac str ui in
+          Cryptokit.xor_string ui 0 u 0 (Bytes.length u);
+          loop (it - 1) ui
+    in
+    loop (iter - 1) ui
+
+  (* The next functions are an implementation of
+        https://www.rfc-editor.org/rfc/rfc5802#page-7 *)
+  let salt_passwd passwd server_first = hi passwd server_first.salt server_first.iter
+
+  let auth_message client_first_message_bare server_first_message client_final_message_without_proof
+      =
+    (* client_first_message_bare = n=$USER,r=$CLIENT_NONCE
+       server_first_message = r=$NONCE,s=$SALT,i=$ITER
+       client_final_message_without_proof = c=$CHANNEL_BINDING,r=$NONCE *)
+    Printf.sprintf
+      "%s,%s,%s"
+      client_first_message_bare
+      server_first_message
+      client_final_message_without_proof
+
+  let client_key salted_passwd = hmac salted_passwd "Client Key"
+  let server_key salted_passwd = hmac salted_passwd "Server Key"
+
+  let server_signature server_key auth_message =
+    hmac server_key auth_message |> Base64.encode_string
+
+  let stored_key client_key =
+    let hash = Cryptokit.Hash.sha256 () in
+    h hash client_key
+
+  let client_proof client_key client_signature =
+    let result = Bytes.of_string client_signature in
+    Cryptokit.xor_string client_key 0 result 0 (Bytes.length result);
+    result |> String.of_bytes |> Base64.encode_string
+
+  (* SCRAM Steps
+
+     Check RFCs 5802 or 7677 to get a better idea on each 
+     exchange format.
+
+     https://www.rfc-editor.org/rfc/rfc5802#section-5
+     https://www.rfc-editor.org/rfc/rfc7677.html#section-3 *)
+  let client_first user passwd =
+    let nonce_len = 16 in
+    let random_data = CCString.init nonce_len (fun _ -> Char.chr @@ Random.int 256) in
+    let nonce = Base64.encode_string random_data in
+    let message = Printf.sprintf "n,,n=%s,r=%s" user nonce in
+    let client_request = { nonce; user; passwd; message } in
+    (client_request, message)
+
+  let parse_server_reply message =
+    (* The first server response looks like this:
+        |----------nonce-----------|
+      r=<client_nonce><server_nonce>,s=<salt>,i=<iter> *)
+    let fields = CCString.split ~by:"," message in
+    let extract prefix =
```

Comment:
I think this part should probably be implemented with pattern matching rather than extract iterating the list each time.  So something that looks more like:

```
match extract fields with
| Some (r, s, i) -> ...
```